### ISSUE TITLE
Build failure due to upstream vendoring

### ISSUE THREAD

**Issue by agy** [Azure/packer-azure#191]
I recently attempted to build packer-azure and failed due to missing dependencies.

The upstream dependency of `azure-sdk-for-go` recently switched to Go vendoring (https://github.com/Azure/azure-sdk-for-go/commit/10a9f00fbc45293d677e2c416285ece01b011ded) resulting in build failures.

Attempting to change the rewritten paths to the new vendored paths fail as Go seems to disallow it.

**Build failure:**
```
root@3f183e751d95:/go/src# go get -v github.com/Azure/packer-azure/packer/plugin/...
github.com/Azure/azure-sdk-for-go (download)
package github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/azure: cannot find package "github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/azure" in any of:
	/usr/local/go/src/github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/azure (from $GOROOT)
	/go/src/github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/azure (from $GOPATH)
package github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/to: cannot find package "github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/to" in any of:
	/usr/local/go/src/github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/to (from $GOROOT)
	/go/src/github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/to (from $GOPATH)
```

**Build failure after attempted path rewrite:**
```
root@3f183e751d95:/go/src# go get -v github.com/Azure/packer-azure/packer/plugin/...
package github.com/Azure/packer-azure/packer/plugin/packer-builder-azure-arm
	imports github.com/Azure/packer-azure/packer/builder/azure/arm
	imports github.com/Azure/azure-sdk-for-go/vendor/github.com/Azure/go-autorest/autorest/azure: use of vendored package not allowed
package github.com/Azure/packer-azure/packer/plugin/packer-builder-azure-arm
	imports github.com/Azure/packer-azure/packer/builder/azure/arm
	imports github.com/Azure/azure-sdk-for-go/vendor/github.com/Azure/go-autorest/autorest/to: use of vendored package not allowed
```

Perhaps it's time start vendoring now that Go seems to have settled on a standard in 1.6?

https://tip.golang.org/cmd/go/#hdr-Vendor_Directories

**Comment by colemickens** [comment]
@paulmey @boumenot How do you feel about adopting Glide? I think that @agy is right. You'll break for any duration between us introducing breaking changes in `go-autorest`, and getting `azure-sdk-for-go` updated to support those changes.

There are a couple ways of alleviating this pain:

1. The most direct: you vendor your dependencies using glide.

2. Ask `azure-sdk-for-go` to check in their vendored dependencies. Then, with Go 1.6 or GO15VENDOREXPERIMENT=1, you would be building with `azure-sdk-for-go's` vendored version of `go-autorest`. (Link to the Vendor rules: https://docs.google.com/document/d/1Bz5-UB7g2uPBdOx-rw5t9MxJwkfpx90cqG9AFL0JAYo/edit#!)

CC: @brendandixon @ahmetalpbalkan

**Comment by boumenot** [comment]
Agreed, @agy is right.  We should fix this.

**Comment by brendandixon** [comment]
@agy @colemickens @boumenot Storing packages within a vendor directory makes sense for the outer most package, but causes problems when sub packages try to do so. We used to store dependent packages in a vendor directory (using Godeps). We backed off (just recently) since it was causing other users of the SDK grief. It's possible that some of that grief was introduced by Godeps (which rewrote the import paths) vs. the model toward which Go is moving.

The problem is that the full import path determines the type and any type sharing between a package and a consumer breaks when they rely on different versions. The ARM packages of the SDK return types defined in go-autorest (e.g., enhanced error types, responses). The consumer must, then, also rely on go-autorest to consume these types. If the SDK uses a different version of go-autorest (drawn from a local vendor directory) than the consumer (pulling from Github or their own vendor directory), compilation breaks.

The only guaranteed solution is for packages to not vendor, but to clearly state their dependencies and for consumers to vendor as needed.
